awx/main/management/commands/cleanup_jobs.py

```diff
@@ -58,27 +62,44 @@ def add_arguments(self, parser):
                             help='Remove workflow jobs')
 
     def cleanup_jobs(self):
-        #jobs_qs = Job.objects.exclude(status__in=('pending', 'running'))
-        #jobs_qs = jobs_qs.filter(created__lte=self.cutoff)
         skipped, deleted = 0, 0
-        jobs = Job.objects.filter(created__lt=self.cutoff)
-        for job in jobs.iterator():
-            job_display = '"%s" (%d host summaries, %d events)' % \
-                          (str(job),
-                           job.job_host_summaries.count(), job.job_events.count())
-            if job.status in ('pending', 'waiting', 'running'):
-                action_text = 'would skip' if self.dry_run else 'skipping'
-                self.logger.debug('%s %s job %s', action_text, job.status, job_display)
-                skipped += 1
-            else:
-                action_text = 'would delete' if self.dry_run else 'deleting'
-                self.logger.info('%s %s', action_text, job_display)
-                if not self.dry_run:
-                    job.delete()
-                deleted += 1
+
+        num_to_delete = Job.objects.filter(created__lt=self.cutoff).count()
+
+        batch_size = 10000
+        num_deleted = 0
+        self.logger.info("deleting total of %d jobs", num_to_delete)
+        while num_deleted < num_to_delete:
+            pk_list = Job.objects.filter(created__lt=self.cutoff)[0:batch_size].values_list('pk')
+            qs_batch = Job.objects.filter(pk__in=pk_list)
+            num_deleted += qs_batch.count()
+            if not self.dry_run:
+                qs_batch.delete()
+            self.logger.info('deleted %d jobs', num_deleted)
 
         skipped += Job.objects.filter(created__gte=self.cutoff).count()
-        return skipped, deleted
+        return skipped, num_deleted
+
+    def cleanup_jobs_fast(self):
+        skipped, deleted = 0, 0
+
+        num_to_delete = Job.objects.filter(created__lt=self.cutoff).count()
+
+        batch_size = 1000000
+        num_deleted = 0
+        self.logger.info("deleting total of %d jobs", num_to_delete)
+        while num_deleted < num_to_delete:
```

Comment:
Related to what @jbradberry mentions about `.delete()` returning a count, another way to do this loop would be instead of doing math on each iteration, just break out of the loop once there's nothing else to delete:

```python
if collector.delete() == 0:
    break
```

Reply:
changed!

Reply:
Yeah, I like this.